File: Python/from_ecote_book/p258_actual_practice_2.py
# 양방향 그래프
# 한 번 이동하는데 시간 1초 걸림
INF = int(1e9)

# n(전체 회사 - 노드의 수), m(경로 - 간선의 수) 입력 받기
n, m = map(int, input().split())

# 연결된 두 회사의 번호 입력 받기
graph = [[INF]*(n+1) for _ in range(n+1)]

# 자기 자신 -> 자기 자신 비용 0으로 초기화
for i in range(1, n+1):
  for j in range(1, n+1):
    if i == j:
      graph[i][j] = 0

# 연결된 a, b 두 회사 정보 입력 받고 서로에게 가는 비용은 1로 초기화
for _ in range(m):
  a, b = map(int, input().split())
  graph[a][b] = 1
  graph[b][a] = 1

# x(최종 목적지)와 k(경유지) 입력 받기
x, k = map(int, input().split())

for k in range(1, n+1):
  for i in range(1, n+1):
    for j in range(1, n+1):
      graph[i][j] = min(graph[i][j], graph[i][k]+graph[k][j])

distance = graph[1][k] + graph[k][x]

if distance >= INF:
  print(-1)
else : print(distance)

File: JavaScript/from_ecote_book/actual_practices/p258_actual_practice_2.js
// 입력 받기
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");
const INF = parseInt(1e9);

// n, m
const [n, m] = input[0].split(" ").map(Number);

const graph = Array.from({ length: n + 1 }, () => Array(n + 1).fill(INF));

// 자기 자신 -> 자기 자신 비용은 0으로 초기화
for (let i = 1; i < n + 1; i++) {
  for (let j = 1; j < n + 1; j++) {
    if (i === j) graph[i][j] = 0;
  }
}

// 연결된 회사 a, b 경로 graph로 정보 입력 받기
for (let i = 1; i < m + 1; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  graph[a][b] = 1;
  graph[b][a] = 1;
}

for (let k = 1; k < n + 1; k++) {
  for (let i = 1; i < n + 1; i++) {
    for (let j = 1; j < n + 1; j++) {
      graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
    }
  }
}

const [x, k] = input[m + 1].split(" ").map(Number);

const distance = graph[1][k] + graph[k][x];

console.log(distance >= INF ? -1 : distance);
